packages/react/src/side-nav/SideNavLinkList.tsx

```diff
@@ -1,51 +1,121 @@
-import { useMemo } from 'react';
-import { SideNavGroup } from './SideNavGroup';
-import { SideNavLink } from './SideNavLink';
+import { useCallback, useMemo } from 'react';
+import { SideNavUnorderedList } from './SideNavUnorderedList';
+import { SideNavListItem } from './SideNavListItem';
 import { SideNavProps } from './SideNav';
 import { hasNestedActiveItem } from './utils';
+import { SideNavLink } from './SideNavLink';
 
 export type SideNavLinkListProps = {
 	activePath: string | undefined;
 	items: SideNavProps['items'];
+	showSubLevel: SideNavProps['showSubLevel'];
 };
 
-export function SideNavLinkList({
-	activePath,
-	items: itemsProp,
-}: SideNavLinkListProps) {
-	const items = useMemo(() => {
-		return itemsProp.map((item) => ({
+type ItemWithIsActive = Omit<SideNavLinkListProps['items'][number], 'items'> & {
+	isActive?: boolean;
+	items?: ItemWithIsActive[];
+};
+
+// Recursively add `isActive` to any sub-level items
+const addIsActive =
+	(activePath: SideNavLinkListProps['activePath']) =>
+	(item: ItemWithIsActive): ItemWithIsActive => {
+		const withNestedActiveItems = hasNestedActiveItem(item.items, activePath);
+		const isCurrentPage = item.href === activePath;
+		const isActive = isCurrentPage || withNestedActiveItems;
+
+		return {
 			...item,
-			isActive:
-				item.href === activePath || hasNestedActiveItem(item.items, activePath),
-		}));
-	}, [activePath, itemsProp]);
+			isActive,
+			items:
+				item.items?.length || isActive
+					? item?.items?.map(addIsActive(activePath))
+					: item.items,
+		};
+	};
```

Comment:
I think this can be simplified and made more performant to by only doing the things necessary at certain points:

```
const addIsActive =
	(activePath: SideNavLinkListProps['activePath']) =>
	(item: ItemWithIsActive): ItemWithIsActive => {
		if (item.href === activePath) {
			return {
				...item,
				isActive: true,
			};
		}

		if (!item.items?.length) {
			return item;
		}

		return {
			...item,
			items: item.items.map(addIsActive(activePath)),
		};
	};
	```

Reply:
This is more performant, for sure. Unless you have hundreds of thousands of items, though, no difference will be noticed by users. I'm also hesitant to implement multiple return statements because these are difficult to compare and often end up further separated as additional requirements are added and devs need to inject more code in between return statements.

I've updated the code to have fewer steps and be more performant, though.

Reply:
I personally don't agree as I think return early is a good pattern that reduces both computational and cognitive overhead (just like you'd return early in a component render where you can't spread stuff around). L31 to L33 are weird with ternaries and setting `item.items` even when it may have been previously non-existant. 

However, after seeing that the original did spreads and stuff, I don't mind too much, and yours works, so it's fine.